### DISCUSSION TITLE
ES6 Express app, using socket.io in other backend files:   'Cannot access 'io' before initialization'

### DISCUSSION THREAD

**Discussion by respectabiggle** [socketio/socket.io#4829] [Q&A] [answered]
> TLDR  When I import { io } from its home file (server.js) into another module, is there a way to asynchronously delay calling io until it loads?

In the simple repo linked below, one module works, but the other module throws an error.

In the module that works, there is some delay (a cron function) before io is called.  Things also work with a setInterval() function.

In the module that does not work, io is employed with no delay, and I get the error  'Cannot access 'io' before initialization'.

I have tried writing the non-working module as an async function ("await io"), but this does not work either.

> note:  The module works if I call its function from server.js.  But I need to be able to call its function from its own file, or from some yet other file in the app.

https://github.com/respectabiggle/socket-test 

Imports

	import { log } 			from 'console'
	import { io }			from './server.js'
	import cron 			from 'node-cron'

   Works

    export const cronTest = function () {

	    let i = 1

	    cron.schedule(`*/1 * * * * *`, async () => {
		    log(i)
		    io.emit('log', `cron: ${i}`)
		    i++
	    })

    }
    cronTest()   	

Does not work

    export const normalTest = function () {

	    let i = 1
	    log(i)
	    io.emit('log', `normal: ${i}`)

    }
    normalTest()							

  Thank you!

**Comment by darrachequesne** [comment, marked as answer]
Hi! I think that's because you have a circular dependency in your code:

```
test cron > server > test normal > server
```

Could you please check?

**Comment by respectabiggle** [comment]
Oh wow yes, you are right.

Thank you for taking the trouble to read my thread!!

I hope you have a great day.
